### DISCUSSION TITLE
🗣️ Webpack improvements

### DISCUSSION THREAD

**Discussion by rigor789** [NativeScript/rfcs#1] [2. Archives]
> **Note:** this document is still a draft and will change
* Goal: Improve usage and maintenance overhead of `@nativescript/webpack`
* Leads: @farfromrefug, @rigor789 

Webpack configs are currently handled by having separate config templates for different project flavors. These templates are copied in the project's root folder in an npm postinstall hook (unless a config already exists). The correct config is determined based on installed dependencies. 

Making changes to the official webpack configs is difficult as the changes usually have to be repeated for each template. After an update, users need to manually update their configs, or alternatively delete their existing ones and install `@nativescript/webpack` again in order to re-generate the config. This is generally not an issue, but many times projects need to customize the configs, making the update process even more difficult.

**Project Overview**

Create a new package in the workspace `packages/webpack5` - it would still be published as `@nativescript/webpack`. The reason is to be able to start from scratch and rebuild it from the ground up with maintenance and better usage in mind - and allow to maintain current configs without interference.

Simplify maintenance by standardizing custom loader implementations, and their location in a `loaders` folder (or where it makes sense).

**Design Goals**
* build an extensible `baseConfig`
* different flavor-specific configs to extend from `baseConfig` (chainable)
* user configs to extend from flavor-specific configs

The `baseConfig` would generally be a regular webpack config that exports a function
```js
module.exports = env => {
  // regular webpack rules that are shared across all flavors
  // workers, entries etc.
  return {
    // config
  }
}
```

A flavor specific config would extend the `baseConfig`:
```js
const baseConfig = require('./baseConfig')

module.exports = env => {
  const config = baseConfig(env)
  
  config.module.rules.push({
    // a flavor specific module rule for example
  });

  return config;
}
```

Users would have a `webpack.config.js` that extends from the flavor config:
```js
const { baseVueConfig } = require('@nativescript/webpack')

module.exports = env => {
  // allows making changes to env prior to calling the base config
  const config = baseVueConfig(env)
  // allows making changes/additions to the base config here
  return config;
}
```

**Comment by jamescodesthings** [comment]
+1 for this, one of the painful parts of the build for me is that any time the template is updated I have to rewrite portions of the new template.

Currently:
- The angular template is not fully ready for worker use, so I have to update the rules as per the `nativescript-worker-loader` guides.
- Aliases: We have a few that have to be pumped in.
- Env & Define: We use a few additional `--env` options to turn on/off features at build time. That are then pushed as vars via DefinePlugin.

If the config was extendable so that it easier to update that would be awesome.

**Comment by NathanaelA** [comment]
I started playing with an idea a week or so ago which allows plugins & apps to modify the config and it worked really well for my initial tests, but I totally agree we need to go to one "base" config as I looked at trying to port it to all of them, and decided that was a chore for another day and that it should just be done when we do a full revamp as the idea needed more fleshing out.  :grinning:   

I'm dropping this out to show the thinking behind the idea; I've not totally sold on doing it this way, but the end result I would like to see plugins and the user be able to easily create customization via a `webpack.include.js` file kinda like the `gradle.include` or `Podfile` does.

Basically in a nutshell the steps I tested was:
1. Use the already included glob() command to find all `webpack.include.js` files, so this step scans _all_ directories and builds a list of all these files.   (Glob is probably the slowest method, but used glob since it is already included in our default install and it is quick to prototype with, quicker way would be to only hit the main directory of each node_module and app_resources folder) --- but the basic idea is that a user can put a `webpack.include.js` in his app_resources folder for his customization's and plugins can include changes if they need in their copies...   
2. Read each one in, and allow them to build a config structure of key/value or key->key->...->value objects as we just take the information exported from it and append it/modify the `customConfigs` values. 
3. This creates a config structure that looks like all the keys and values that we use; so for example lets take appComponents:

```
const customConfigs = {appComponents: env.appComponents || [], ... all the config default settings for every key value we use ...};

// Process all custom configs...
for (file in listOfConfigs) { processConfig(customConfigs, file); }

// Later in webpack; 
const appComponents = customConfigs.appComponents;   // So now appComponents has both env values & custom config values
appComponents.push(...[
        "@nativescript/core/ui/frame",
        "@nativescript/core/ui/frame/activity",
    ]);
```
 
---

After building it this way the  problem with this way; is that the custom configs couldn't make any changes to defaults(for example, it couldn't remove/replace `@nativescript/core/ui/frame` from the appComponents if it needed to, so I think it might be better to do it:

1. Find all webpack.includes, these includes export one (or optionally a 2nd) function, and a optional sortValue; 

2. Sort them based on if the sortValue exported.  (no sort value = 0, so by default ordered to running first, as they are probably just adding things like a copy rule or something simple that has no need to be ran last)

3. Run the `setup` function in each webpack.include with the current defaults and structure of rules (i.e. all the values that would be passed into all the objects when finally creating the "config" object, so it would be the `customConfig` as above..)     -- I believe 99% of the changes would probably be handled in this function as you would add new copy rules, change default values, create your overrides for values that will be passed into creating the config.    All default values set in the config would be exposed here and then this customConfig would used to create the final config that is passed to webpack.
(run step 3 for each include file)

4. Run the `process` function in each webpack.include with final webpack config as built (to allow the include file to make any extra decisions based on the config as it currently stands and make any changes it needs to) -- I don't see this being used very much, but I would rather expose this ability because maybe instead of using Terser, they want to use Uglify...
(Run step 4 for each include file)

**Comment by farfromrefug** [comment]
I personally was thinking of something  a lot more strait forward in the sense of requiring a lot let work from us. 
Basically the user calls tns which calls webpack (CLI or Api?) Using it s defined webpack config (default generated or custom one defined in config). 
That webpack config,as said @rigor789 , calls a flavor one 
That flavor one calls the basic default N one.
Now here is the magic part I am thinking about. I would like the flavor and the basic one to call N hooks on the config they generate! That way plugin can anything the webpack way!
Now is it doable? @rigor789 how did you manage to inject the "object" which handles the nativescript config in webpack ? You showed me the api which was able to read from the ts config file. Could we do it that way?

**Comment by rigor789** [comment]
@farfromrefug an example is in the PR: https://github.com/NativeScript/nativescript-cli/pull/5381

We could probably call into hooks with that.

I would go for explicit inclusion: plugins can export a function users need to call from their webpack configs:

```js
const { baseVueConfig } = require('@nativescript/webpack')
const setupSomePlugin = require('someplugin/setup') // naming can be a convention we decide

module.exports = env => {
  // allows making changes to env prior to calling the base config
  const config = baseVueConfig(env)

  // apply rules from the plugin by calling it's exported setup function
  // it's free to do any modifications to the config
  // optionally - it could be called prior to calling the base config, since these would be plain functions
  setupSomePlugin(env, config);

  // allows making changes/additions to the base config here
  return config;
}
```

Magically including webpack configs could lead to hard to debug/track down build issues in my opinion.

**Comment by farfromrefug** [comment]
I dont want to do it magically. I am thinking more in the lines of doing it in the right order:
Base config
Bade config hooks
Flavor config
Flavor hooks
User config
And it would be a call explicit but not at the user level. And we could imagine an env option  ofr something to disiable those so that the user still gets the opportunity to disable them.

**Comment by NathanaelA** [comment]
@farfromrefug - The design I listed would handle that exact scenario without complicating the actual CLI side of things and I think that would be an immense over-complication to make the CLI aware of webpack values beyond calling it with the few parameters it sends right now.    

I would agree that the path to the nativescript.config file path should also passed to the webpack tooling as another parameter so it could require it and use any values set it in.      I would rather make the CLI a lot more agnostic so that we can swap in other build tools later, and/or let the community try to create other bundler tool integrations.    Long term Webpack should be a easily swap-able piece and not a requirement.      I would argue for, that we should add a value to the nativescript.config file for what bundler to call via the cli (unset = webpack) and make it call the bundler with the path to the nativescript.confg and all parameters passed to the cli and any others we generate during the setup of the cli.  

I strongly believe the cli & webpack should NOT be dependent upon each other beyond the cli calling a bundler then calling the native build tools.   It should be divorced as much as possible from both...

**Comment by NathanaelA** [comment]
@rigor789 -  as for risk, it would add (imho) less risk to the tooling than the current `Podfile`, `build.xcconfig` & the `include.gradle` files we already support per plugin -- those can break the tooling just as easily and are a lot harder to trace down.   So I hope you aren't also advocating rolling back that automagical support too as it would make plugins a lot more painful to use for users.    I know for sure I would be very unhappy if we ditched any of those abilities and made the tooling require a user to manually copy/paste the podfile/gradle code we would then have to include in our readme for them to manually put in a single manual app_resource only Podfile/gradle file location ...   

The number of support issues I've seen because people can't read the instructions on how to add a change to the webpack copy rules to simply copy their sqlite files in the build is something I would like to easily "automagically" solve with a rework so it is no longer a issue and so adding the ns-sqlite plugin would automagically allow `**/*.sqlite` files to be included in the copy files list.   :grinning:    

---


The three use cases I know I would use immediately if this was supported; and all of them really should be automagicly done by installing that plugin...

- Adding new copy rules and ignore rules (Sqlite plugin, the NS Automated Test App)
- Modifying different rules & switching engines to eliminate certain data leaks (Encryption plugin)
- Ability to pre-run a ionic build, then setting up copy rules to copy the ionic dist folder into the final assets folder so it is included in the NS app as copied files.  (NS-Ionic plugin)

Which actually brings up that the override system needs to be async, so that it can await on external changes (i.e external tooling builds, and external data lookups, etc)

---

 **I would like plugins to be just as much a first class citizen to Webpack as they currently are for Gradle and Cocoapods.**

In fact I would strongly argue that this "plugin" mechanism works exactly the exact same way for the flavors since flavors really are just plugins.  The flavors `include.webpack` file ships with the flavor as part of the flavor and is maintained with and by the flavor author, and the main webpack file would just include it just like it would any other plugin, and the flavor would override the things it needs to...   

Webpack Flavor support should **NOT** be treated any differently than any other plugin, keep all the tooling 100% consistent.  
Maybe look at it that way would make more sense to you Igor...  :grinning:   

---
So maybe:

Virtually empty USER Webpack config is shipped with the app (or generated if empty); 
```
// The require statement will also trigger looking for all the include.webpack files, and parsing them and 
// running the setup on them to generate a final Webpack.settings structure all ready to be used.
const Webpack = require("@nativescript/webpack");

// User can now optionally modify any Webpack.settings items as any final overrides should be at this, the user level.
// i.e. something like: webpack.settings.myAwesomeValue = "Awesome";

// This will build the final webpack config structure using all the settings 
// and then run each of the `include.webpack` process functions
// finally returning the complete `config`
const config = await Webpack.generateConfig(webpack.settings);  

// Again user can optionally modify any actual final config items (like say replace Terser with Uglify, or any other config items)
// i.e. something like: config.someWebpack.config.item = new Webpack.CopyPlugin(blah);

// And then actually run webpack with our final config...
Wepack.run(config);
````

The @nativescript/webpack index.js would be basically:
```
// all the BASE default settings, including anything loaded from nativescript.config.js and any env values passed
const settings = { nativescript: {nsConfigValues}, env: { envValues}, someSetting: ... }

// Scan for all include.webpack,  load them in and sort them...
// Internally scanForIncludes would have try/catch around each load to catch any errors and report them
const includes = scanForIncludes(...);    

// Process each Include the "includes" array will already be in the proper sort order...
// processInclude also has a try/catch internally to catch any errors and log them...
foreach (include in includes) { 
   processInclude (settings, include); 
}

exports.settings = settings;
exports.generateConfig = async (rules) => { /* use rule settings to generate config, then call each include.process function with it */ }
exports.run = (config) => { /* run webpack w/ config */ }
```



The per plugin/flavor `include.webpack` would be simply
```
// Required - Version this include subscribes too...  To allow future proofing very easily.  
export.version = 1;

/* optional, default = none */ 
export.setup = (settings) => { 
/* make any changes to settings based on settings passed in and anything it knows it needs to do */
}

/* optional, default = none */
export.process = async (config) => {
   /* make any changes to the final config object */
};

/* optional, default = 0 */
export.sortValue = 500;
```

**Comment by farfromrefug** [comment]
@NathanaelA I agree with you on most things
I just don't see any need for your process includes stuff nor any logic our our side. Everything can be simply done within the normal webpack config files. 
This is what we did for svelte for quite some time and it work well. 
My point is the less special handling we add the less we have to maintain. And users can refer to webpack doc. We then don't need to write any.
@rigor789 looked at webpack 5 and I don't see any blocking thing there. And damn I want it for tree shaking !

**Comment by rigor789** [comment]
> I strongly believe the cli & webpack should NOT be dependent upon each other beyond the cli calling a bundler then calling the native build tools. It should be divorced as much as possible from both...

@NathanaelA That is one of my goals as well - however the CLI is already fairly coupled to webpack, and decoupling it is out of scope for this project. I would love to revisit it at a later point though, especially now that the bundler space is very much improving and alive (vite, esbuild, parcel etc.). 

> So I hope you aren't also advocating rolling back that automagical support too as it would make plugins a lot more painful to use for users. 

That's just twisting my words. That's not what I suggested...

> The flavors include.webpack file ships with the flavor as part of the flavor and is maintained with and by the flavor author, and the main webpack file would just include it just like it would any other plugin

I feel like managing the flavor specific configuration in the `@nativescript/webpack` plugin is easier to maintain - no need to over-complicate things, that's what leads to unmaintainable mess. 

As for the example usage you are suggesting - it's beyond what webpack api's provide, introducing a new "DSL" and complexity is not what we should be focusing on, but we can definitely iterate on the idea!

Auto-discovery of webpack configs can be much simpler, and **easy to opt-out.** Plugins could ship a configuration file in a standard location, with the standard webpack syntax:
```js
// the file-name is debatable, since non-ns plugins could potentially include a webpack config if not packaged correctly
// node_modules/nativescript-plugin-example/webpack.config.js

// the only deviation from a standard webpack config would be the second parameter
module.exports = (env, config) => {
  // the plugin is free to make changes to the config
  config.plugins.push(new CopyWebpackPlugin({ /* ... */ }))
}
```

The project configuration would then explicitly call a `applyPluginConfigs` (the name is a placeholder - i'm sure it can be improved):
```js
const { baseVueConfig, applyPluginConfigs } = require('@nativescript/webpack')

module.exports = env => {
  // allows making changes to env prior to calling the base config
  const config = baseVueConfig(env)

  // apply rules from all plugins that have customizations
  applyPluginConfigs(env, config)

  // optionally, the user can be specific by passing the plugin names
  applyPluginConfigs(env, config, {
     // this syntax would even allow passing parameters to plugins
    'nativescript-plugin-example': {
      someOption: true,
      maxSomething: 200
    }
  })
  
  // allows making changes/additions to the base config here

  return config;
}
```

The syntax and API is up for discussion, there are missing scenarios that we can solve by providing different methods:
* apply all plugins, but allow specifying options for specific plugins.

This approach would get us closer to best-of-both-worlds.

**Comment by NathanaelA** [comment]
>> I strongly believe the cli & webpack should NOT be dependent upon each other beyond the cli calling a bundler then calling the native build tools. It should be divorced as much as possible from both...

> @NathanaelA That is one of my goals as well - however the CLI is already fairly coupled to webpack, and decoupling it is out of scope for this project. I would love to revisit it at a later point though, especially now that the bundler space is very much improving and alive (vite, esbuild, parcel etc.).

I am aware that they are still coupled; but the webpack config can be setup so that it doesn't care about the CLI.   So I was trying to say don't couple it more with the CLI, we want to eliminate as much coupling as possible.    (Also agreed this is out of scope -- but was just trying to warn against going towards more coupling it direction, and not saying we need to do this now.  :grinning: )

---

>>> Magically including webpack configs could lead to hard to debug/track down build issues in my opinion.

>> So I hope you aren't also advocating rolling back that automagical support (podfile/gradle/xcconfig) too as it would make plugins a lot more painful to use for users.

> That's just twisting my words. That's not what I suggested...

No, I was taking your prior (quoted above) statement to its logical conclusion and trying to show how this position is the wrong position for the end users.   If your "fear" is that is could create broken builds then by all means the other three files that we let plugins change should also be removed as they can cause the exact same issues and are way **harder** to track down.       I was hoping that showing the absurdity of that position would help you realize that automagical is what we **do** want, and we do not want anything manual... :grinning:     Now, maybe an flag that can list all included webpack configs or maybe with the verbose we list every included webpack config.   But I strongly believe in automagical as much as possible, this has to be as simple stupid for the end user developer as possible.

---

First, when I use plugins it is interchangeable with flavor; Since a flavor is just a plugin that adds a different template binding/ building and still uses the core to do its work.   So it simplifies me from having to say plugins/flavor everywhere.   :grinning:

> I feel like managing the flavor specific configuration in the @nativescript/webpack plugin is easier to maintain - no need to over-complicate things, that's what leads to unmaintainable mess.

I would totally disagree with you on this one -- the current mess is because the NS team has been maintaining them.   The people who maintain the plugin would have the best idea what `extra` support they need beyond the default config.   We at the TSC just have to make sure that the base webpack config is easily modifiable by any plugins and be responsive to make additional changes in the base if they come across some feature that a flavor/plugin needs that should be in base.   I know for sure I don't know what needs to be added to make Angular, Svelte or even Vue in the webpack config to make them work properly, etc...  So putting the maintenance of `extra` options where I believe it belongs (on the plugin author who knows what they need) seems like the best place to put it. This allows new flavors to be EASY created without having to deal with the TSC to get any "official" webpack file added and allows any existing flavor to easily be UPDATED again without getting the TSC involved!    Again the whole idea here is since flavors are a plugin, maintaining flavors and plugins should not have to have the TSC get involved and they should be able to push to npm **everything** they need at any point without any TSC involvement.   I strongly believe we should NO longer maintain any flavor webpacks with the webpack plugin; but hand them back to their respective authors.    I do believe we should make the new version for them initially, so they don't have to transition it, but hand them back to be included in their flavor/plugin.


> As for the example usage you are suggesting - it's beyond what webpack api's provide, introducing a new "DSL" and complexity is not what we should be focusing on, but we can definitely iterate on the idea!

This isn't complicated at all, and no real new DSL needed, just a simple settings object (like the nsconfig file); I had a working system in less than an hour.   I didn't have nearly as `clean` as I think we need it to be -- to make it as much future proof as possible; but I had the TypeScript flavor fully working with UNLIMITED external webpack.include.js files to add multiple rules to it, so that the Automated App could work 100% and Sqlite and NS-Ionic plugins could work properly all from the same "master" webpack file, but all three of those had totally different requirements in what they needed from webpack...

---

As for your suggestion; the biggest issue with it (as it was my first test when I was trying ideas the other day  :grinning:) is that without some of the settings values (that are used to build the config) it is hard to retrieve existing values and/or create the changes needed. 

   So here are some use cases that need to be solved and we want to do the heavy lifting so that they are trivial for the plugins to use -- I believe they are a good representation of a number of cases that we will run into; as they are real cases that I have ran into:

1. The CopyWebpackPlugin rule, how does all the additional plugins include files get the current `copyIgnore` value so that the newly added `CopyWebpackPlugin` will respect what is supposed to be ignored? 

2.  On the opposite side; even more important how does a plugin undate the copyIgnore rule so ALL existing and future `CopyWebpackPlugin` will ignore all the path's now set, to be ignored...

3. How does one override an existing loader get its existing values so it can add the additional new options.   So for example; say I wanted to replace the `loader: "@nativescript/webpack/bundle-config-loader"` or `loader: "ts-loader"` with additional config values; how do I get the values already passed in so that when I replace the loader I can use not only the current values but my additional config items that I also need added to loader.    Please note, that base TS config could add ts-loader, then Angular could create its own variation of ts-loader, and then a plugin could have its variation because it needs some extra rules applied to the TS-load.   All variation versions rules need to be accessible so they can be combined.    So how does Angular get the base TS config, and then the plugin get the Angular changes that Angular applied to the base.   We can't just have the plugin author attempt to re-compute the values based on the base config --  because it has NO idea what other changes may have occurred between the "base" and itself when it attempted to re-created the loader because of the additional rules it needs.

4. How do I run a external process, and await it to be completed before I continue the webpack building as webpack depends on this process being finished?

5. How do we guarantee some plugins have their webpack config applied first or last?

Every one of the above questions has been generalized but is a real issue with at least one plugin I know of in the community.

---

So hence my suggestion of a settings phase, and then a config phase, as it solves every single question above.  :grinning:   

 I am not opposed to doing it differently at all; but I do feel strongly that the questions above need to have a **solid** method to be handled **automatically** in whatever direction we proceed forward.   The plugins need to be a full first class citizen.

If you would like I can create a PR with a base system like I am proposing --- unless you have a solid way to make everyone of these questions be trivial to handle in the PR that you have submitted by a plugin.

**Comment by rigor789** [comment]
The details aren't clear yet - as we are still moving things around and figuring stuff out, but the progress on the PR looks like it will be able to solve all the issues you mentioned, as we opted to use webpack-chain which gives us the flexibility to modify anything in the config.

A very early example of it is something like: tap into the existing `ts-loader` options, and just add `appendTsSuffixTo: [/\.vue$/],`. (we probably want to merge in this case - since there may already be values in  `appendTsSuffixTo`)
```js
// set up ts support in vue files
config.module
  .rule('ts')
  .use('ts-loader')
  .loader('ts-loader')
  .tap((options) => {
    return {
      ...options,
      appendTsSuffixTo: [/\.vue$/],
    };
  });
```

How the `webpack-chain` configs can be exposed to the user - we are still testing different approaches to make it as straight forward as possible, for both users and plugin authors.

**Comment by NathanaelA** [comment]
Very nice, AWESOME Find!     I hadn't seen [Webpack chain](https://github.com/neutrinojs/webpack-chain); that looks exactly like what I wanted for the `Settings`  -- it does actually look much more powerful and could handle everything we need.     I haven't investigated it very deeply; but has this been merged with your PR; I'd like to play with it...

**Comment by rigor789** [comment]
@NathanaelA Yes, it is in the PR, but far from something useful - we only just started: 
https://github.com/NativeScript/NativeScript/blob/feat/webpack5/packages/webpack5/src/configuration/vue.ts 

We set up jest for testing, here's what the vue stuff looks so far (this isn't the actual config, rather a human-readable version from webpack-chain using `toString()` - pretty cool format for debugging, it adds comments to each block, how you may call it from webpack-chain): 
https://github.com/NativeScript/NativeScript/blob/feat/webpack5/packages/webpack5/__tests__/configuration/__snapshots__/vue.spec.ts.snap

It builds on top of `base` (which needs lots more work to be something we can use). No loaders or plugins have been ported into the new package yet, as the goal is to only bring things we actually need - there's a lot of legacy code in the current webpack package.

**Comment by shirakaba** [comment]
Very curious about webpack-chain. Looks nifty; might do the trick.

I’ve felt very limited by straight up extension of Webpack configs (how we’re doing it right now). Hacking around with plugins.find() and plugins.splice() has been very messy. And defining extra WebpackDefinePlugin definitions wasn’t fun, either.

Worst of all has been trying to roll back CopyWebpackPlugin because it is invoked with a new (and mutually exclusive) API in the base config, yet there’s no way to inherit it and swap it out without instantiating an instance of CopyWebpackPlugin (which inevitably causes a crash if you’ve deliberately installed the older version of the plugin, due to the incompatible construction args).

I couldn’t think of an elegant answer to the frictions I encountered (though a few user- and dev-unfriendly options came to mind), so I’m very interested to see what options webpack-chain opens up through its builder pattern. Good find, @rigor789!

**Comment by edusperoni** [comment]
I suggest we take a look at https://github.com/GoogleChromeLabs/worker-plugin. The current worker-loader implementation is based on https://github.com/webpack-contrib/worker-loader

Mainly we don't need stuff like `import Worker from 'worker-loader!./Worker.js';`  and it already has support for reusing plugins (https://github.com/GoogleChromeLabs/worker-plugin#plugins-array) which was something I had to add to ensure compatibility with Angular (https://github.com/NativeScript/worker-loader/pull/39)

**Comment by rigor789** [comment]
@edusperoni seems like Webpack 5 supports workers out of the box, with a slightly different syntax:

```js
const worker = new Worker(new URL('./worker.js', import.meta.url));
```

I've tested the worker-plugin and it doesn't work with webpack5, and the built-in way is mentioned in their docs:
> Compatibility Note: Webpack 5 now includes worker bundling. It uses a slightly different syntax:
    new Worker(new URL("./my_worker.js", import.meta.url))

This still needs testing - or perhaps if the worker-plugin releases a new version that just converts the syntax to the webpack syntax, that would be pretty nice!

**Comment by rigor789** [comment]
Closing in favor of the pending RFC: #2 

Please continue discussion under the RFC!
